File: typescript/examples/telemetry/04_runtime_reconfigure.ts
#!/usr/bin/env npx tsx
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of Provide Telemetry.

/**
 * 🔄 Runtime reconfiguration — hot-swap config without restart.
 *
 * Demonstrates:
 * - getRuntimeConfig / updateRuntimeConfig for hot updates
 * - reconfigureTelemetry for provider-safe reconfiguration
 * - reloadRuntimeFromEnv to re-read environment variables
 * - provider-changing OTEL fields requiring restart after live registration
 *
 * Run: npx tsx examples/telemetry/04_runtime_reconfigure.ts
 */

import {
  event,
  getHealthSnapshot,
  getLogger,
  getRuntimeConfig,
  reconfigureTelemetry,
  reloadRuntimeFromEnv,
  setupTelemetry,
  shutdownTelemetry,
  updateRuntimeConfig,
} from '../../src/index.js';

async function main(): Promise<void> {
  console.log('🔄 Runtime Reconfiguration Demo\n');

  setupTelemetry({ serviceName: 'provide-telemetry-examples', logLevel: 'info', consoleOutput: false });
  const log = getLogger('examples.runtime');

  // ── 📊 Inspect current config ────────────────────────
  const cfgBefore = getRuntimeConfig();
  console.log(`📊 Before: serviceName=${cfgBefore.serviceName}  logLevel=${cfgBefore.logLevel}`);

  log.info({ ...event('example', 'runtime', 'before') });

  // ── 🔧 Hot-swap log level ─────────────────────────────
  console.log('\n🔧 Hot-swapping logLevel to warn...');
  updateRuntimeConfig({ logLevel: 'warn' });
  const cfgAfter = getRuntimeConfig();
  console.log(`  ✅ After update: logLevel=${cfgAfter.logLevel}`);

  log.info({ ...event('example', 'runtime', 'dropped') }); // suppressed at warn level

  // ── ♻️ Non-breaking reconfigure ──────────────────────
  console.log('\n♻️  reconfigureTelemetry() — safe reconfigure (no provider change)...');
  reconfigureTelemetry({ logLevel: 'info', serviceName: 'provide-telemetry-examples-v2' });
  const cfgRestarted = getRuntimeConfig();
  console.log(`  ✅ Reconfigured: serviceName=${cfgRestarted.serviceName}  logLevel=${cfgRestarted.logLevel}`);

  log.info({ ...event('example', 'runtime', 'reconfigured') });

  const healthAfter = getHealthSnapshot();
  console.log(`  📊 Health after reconfigure: logsDropped=${healthAfter.logsDropped} exportFailuresLogs=${healthAfter.exportFailuresLogs}`);

  // ── 🌍 Reload from environment ───────────────────────
  console.log('\n🌍 reloadRuntimeFromEnv() — re-reads process.env hot fields only...');
  reloadRuntimeFromEnv();
  const cfgReloaded = getRuntimeConfig();
  console.log(`  ✅ Reloaded: logLevel=${cfgReloaded.logLevel}`);

  // ── 🚫 Provider-changing fields after live registration ───────────────
  console.log('\n🚫 Provider-changing fields are rejected after OTEL providers are live.');
  console.log(
    '  Reconfigure with otelEnabled / otlpEndpoint / otlpHeaders only before registerOtelProviders(),',
  );
  console.log('  or restart the process and call setupTelemetry() with the new provider config.');

  console.log('\n🏁 Done!');
  await shutdownTelemetry();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/telemetry/04_runtime_reconfigure.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (C) 2026 MindTenet LLC
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

from __future__ import annotations

from provide.telemetry import (
    get_health_snapshot,
    get_logger,
    get_runtime_config,
    setup_telemetry,
    shutdown_telemetry,
    update_runtime_config,
)
from provide.telemetry.config import TelemetryConfig


def main() -> None:
    setup_telemetry()
    log = get_logger("examples.runtime")

    cfg_before = get_runtime_config()
    print({"before_logs_rate": cfg_before.sampling.logs_rate})

    log.info("example.runtime.before")

    # ── 🔧 Hot-swap sampling rate to 0% ──────────────────
    print("\n🔧 Hot-swapping sampling rate to 0%...")
    cfg_after = TelemetryConfig.from_env({"PROVIDE_SAMPLING_LOGS_RATE": "0.0"})
    updated = update_runtime_config(cfg_after)
    print(f"  ✅ After update: logs_rate={updated.sampling.logs_rate}")

    log.info("example.runtime.dropped")

    snapshot = get_health_snapshot()
    print(
        {
            "after_logs_rate": get_runtime_config().sampling.logs_rate,
            "dropped_logs": snapshot.dropped_logs,
        }
    )

    # ── ♻️ Full provider restart via reconfigure ─────────
    print("\n♻️  reconfigure_telemetry() — full shutdown+setup cycle...")
    cfg_restarted = reconfigure_telemetry(TelemetryConfig.from_env({"PROVIDE_SAMPLING_LOGS_RATE": "1.0"}))
    print(f"  ✅ Restarted: logs_rate={cfg_restarted.sampling.logs_rate}")

    log.info("example.runtime.restarted")

    # ── 🌍 Reload from environment ───────────────────────
    print("\n🌍 reload_runtime_from_env() — re-reads os.environ...")
    reloaded = reload_runtime_from_env()
    print(f"  ✅ Reloaded: logs_rate={reloaded.sampling.logs_rate}")

    print("\n🏁 Done!")
    shutdown_telemetry()


if __name__ == "__main__":
    main()
